extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/CdiAwareJob.java

```diff
@@ -19,36 +19,38 @@
  */
 class CdiAwareJob implements InterruptableJob {
 
-    private final Instance<? extends Job> jobInstance;
+    private final Instance<? extends Job> instance;
+    // keep a contextual reference so that we avoid race condition where interrupt() is invoked after execute() finished
+    private Job contextualReference;
+    private Instance.Handle<? extends Job> handle;
 
-    public CdiAwareJob(Instance<? extends Job> jobInstance) {
-        this.jobInstance = jobInstance;
+    public CdiAwareJob(Instance<? extends Job> instance) {
+        this.instance = instance;
     }
 
     @Override
     public void execute(JobExecutionContext context) throws JobExecutionException {
-        Instance.Handle<? extends Job> handle = jobInstance.getHandle();
+        if (handle == null) {
+            this.handle = instance.getHandle();
+            this.contextualReference = handle.get();
+        }
         try {
-            handle.get().execute(context);
+            contextualReference.execute(context);
         } finally {
             if (handle.getBean().getScope().equals(Dependent.class)) {
                 handle.destroy();
+                // needed for jobs that can re-fire; we always want a new dependent bean there
+                // intentionally not cleaning up bean reference as interrupt() can occur after execute()
+                handle = null;
             }
         }
     }
 
     @Override
     public void interrupt() throws UnableToInterruptJobException {
-        Instance.Handle<? extends Job> handle = jobInstance.getHandle();
         // delegate if possible; throw an exception in other cases
         if (InterruptableJob.class.isAssignableFrom(handle.getBean().getBeanClass())) {
-            try {
-                ((InterruptableJob) handle.get()).interrupt();
-            } finally {
-                if (handle.getBean().getScope().equals(Dependent.class)) {
-                    handle.destroy();
-                }
-            }
+            ((InterruptableJob) contextualReference).interrupt();
```

Comment:
Approximately the same rare race condition: the job can be interrupted before `contextualReference` is set...
* I suppose custom `interrupt()` should be called anyway, so need to initialize bean.
* Need not destroy the bean because `execute()` should be run on the same instance.
* The only left is to believe that `execute()` will really be called to have a chance to destroy the bean...

Some alternative thought about `interrupt()` after `execute()`.

Since CustomInterruptableJob.execute(context) is already finished in this case, there is seemingly nothing that can control whether CustomInterruptableJob.interrupt() will be called or not. I mean, just a moment later the job will leave the list of currently running job, `Scheduler.interrupt()` will not find it and return `false`.

If considering this as an undefined behavior, one could stick to one more option - always return `false`, but this requires support by Quartz inside [QuartzScheduler.interrupt()](https://github.com/quartz-scheduler/quartz/blob/a5c4d27e963f51097f9b2777489d310a88897ca4/quartz/src/main/java/org/quartz/core/QuartzScheduler.java#L2329-L2330), like this:

```java
if (job instanceof InterruptableJob) {
    try {
        ((InterruptableJob)job).interrupt();
        interrupted = true;
    } catch (JobWasNotReallyInterrupted) {
        interrupted = false;
    }
}
```
and raise JobWasNotReallyInterrupted in CdiAwareJob if the job is (being) destroyed.

Reply:
> Approximately the same rare race condition: the job can be interrupted before contextualReference is set...

Hm, I was trying to avoid using locks here as it would affect all the jobs and make them more expensive for the sake of a corner case...
We could initialize the first state in constructor which would eliminate this issue for anything but re-firing jobs where this slim chance still exists.

Reply:
> I was trying to avoid using locks here as it would affect all the jobs and make them more expensive

Not dramatically more expensive as locks are already used, e.g. in LazyInstanceHandle.

> We could initialize the first state in constructor

You mean to call `Instance.getHandle()` inside constructor? This will in particular rely on if `execute()` is called eventually to be able to destroy it. I am not sure if this is always the case.

Reply:
> EDIT the same logic for not yet started to execute() jobs.

I am not sure that's the case, quartz wouldn't find that job as running to invoke interrupt on it, would it?

Reply:
As far as I see from [JobRunShell.run()](https://github.com/quartz-scheduler/quartz/blob/a5c4d27e963f51097f9b2777489d310a88897ca4/quartz/src/main/java/org/quartz/core/JobRunShell.java#L150), it
1. Calls notifyListenersBeginning() that in particular registers the job as running.
2. Calls execute()
3. Calls notifyJobListenersComplete() that in particular unregisters the job.

The problem in particular in what should happen with `interrupt()` in periods between 1 and 2, and between 2 and 3 (plus some time after 3, because Quartz iterates over [copy of running jobs](https://github.com/quartz-scheduler/quartz/blob/a5c4d27e963f51097f9b2777489d310a88897ca4/quartz/src/main/java/org/quartz/core/QuartzScheduler.java#L2317) out of lock to find the job for interrupting).

In [comment](https://github.com/quarkusio/quarkus/pull/42096#discussion_r1689795106) I thought about if behavior in these periods can be considered undefined, so that it would be possible to choose the mentioned specific one.